src/mame/machine/dc_g2if.cpp

```diff
@@ -0,0 +1,538 @@
+//#define LOG_OUTPUT_STREAM std::cout
+#include "logmacro.h"
+
+#define LOGWARN(...)      LOGMASKED(LOG_WARN, __VA_ARGS__)
+#define LOGDMA(...)       LOGMASKED(LOG_DMA, __VA_ARGS__)
+#define LOGHWTRIG(...)    LOGMASKED(LOG_HWTRIG, __VA_ARGS__)
+#define LOGDMAEND(...)    LOGMASKED(LOG_DMAEND, __VA_ARGS__)
+#define LOGILLEGAL(...)   LOGMASKED(LOG_ILLEGAL, __VA_ARGS__)
+
+//**************************************************************************
+//  GLOBAL VARIABLES
+//**************************************************************************
+
+
+// device type definition
+DEFINE_DEVICE_TYPE(DC_G2IF, dc_g2if_device, "dc_g2if", "Sega Dreamcast G2 I/F System Bus")
+
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+
+//-------------------------------------------------
+//  dc_g2if_device - constructor
+//-------------------------------------------------
+
+
+dc_g2if_device::dc_g2if_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, DC_G2IF, tag, owner, clock)
+	, m_host_space(*this, finder_base::DUMMY_TAG, -1)
+	, m_int_w(*this)
+	, m_error_ia_w(*this)
+	, m_error_ov_w(*this)
+{
+}
+
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+
+void dc_g2if_device::device_start()
+{
+	for (int i = 0; i < 4; i++)
+	{
+		m_dma[i].end_timer = timer_alloc(i);
+	}
+
+	m_int_w.resolve();
+	m_error_ia_w.resolve();
+	m_error_ov_w.resolve();
+
+	save_item(STRUCT_MEMBER(m_dma, g2_addr));
+	save_item(STRUCT_MEMBER(m_dma, root_addr));
+	save_item(STRUCT_MEMBER(m_dma, len));
+	save_item(STRUCT_MEMBER(m_dma, size));
+	save_item(STRUCT_MEMBER(m_dma, mode));
+	save_item(STRUCT_MEMBER(m_dma, dir));
+	save_item(STRUCT_MEMBER(m_dma, enable));
+	save_item(STRUCT_MEMBER(m_dma, in_progress));
+	save_item(STRUCT_MEMBER(m_dma, start));
+	save_item(STRUCT_MEMBER(m_dma, tsel));
+	save_item(STRUCT_MEMBER(m_dma, hw_trigger));
+	save_item(NAME(m_g2apro.top_addr));
+	save_item(NAME(m_g2apro.bottom_addr));
+}
+
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+
+void dc_g2if_device::device_reset()
+{
+	for (int ch = 0; ch < 4; ch ++)
+	{
+		m_dma[ch].g2_addr = 0;
+		m_dma[ch].root_addr = 0;
+		m_dma[ch].len = 0;
+		m_dma[ch].size = 0;
+		m_dma[ch].mode = false;
+		m_dma[ch].dir = false;
+		m_dma[ch].enable = false;
+		m_dma[ch].in_progress = false;
+		m_dma[ch].start = false;
+		m_dma[ch].tsel = 0;
+		m_dma[ch].hw_trigger = false;
+		m_dma[ch].end_timer->adjust(attotime::never);
+	}
+}
+
+void dc_g2if_device::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	u8 channel = (u8)id;
+	bool dma_result = (bool)(param == 1);
```

Comment:
The cast to `bool` here is pointless – a comparison produces a `bool` result directly.